### ISSUE TITLE
Implement Δ-relative vigilance anchor for real embedding manifolds

### ISSUE THREAD

**Issue by samscarrow** [git-scarrow/fast-associative-memory#73]
## Context
PR #72 closed #71 by validating the ρ(t) onset criteria on real embeddings.
Result: the Δ-relative blend criterion transfers with quantified offset, but absolute ρ thresholds do not transfer. In the real all-MiniLM-L6-v2 run, within-class Δ≈0.44, blend onset was already left-censored at epoch 0 around ρ≈0.242, and max ρ under heavy text-level contraction plateaued around 0.481. The fixed v_ceiling=0.95 chimera criterion is therefore unreachable on this real manifold.

## Goal
Implement and evaluate a relative / percentile dynamic-vigilance anchor that calibrates to the live embedding manifold rather than fixed absolute cosine thresholds.

## Proposed approach
Add a new vigilance policy, separate from the existing margin-based `DynamicVigilance`, that estimates one or more live distribution signals:
- within-class Δ
- off-class ρ distribution
- p90/p95 off-class similarity
- Δ−ρ margin
- off-class softmax vote mass

Then set effective vigilance from a relative rule, for example:
```python
v_eff = clamp(rho_p95 + margin_guard, v_floor, v_ceiling)
```
or:
```python
v_eff = clamp(delta_live - target_blend_margin, v_floor, v_ceiling)
```

The percentile form is preferred as the starting point: the real failure mode in #72 is many moderately similar off-class neighbors aggregating vote mass, not one clean nearest competitor.

## Evaluation
Reuse `benchmarks/probe_contraction.py` in both modes:
- synthetic baseline
- real 20NG + all-MiniLM-L6-v2 transfer path

Compare current `DynamicVigilance` vs the new relative policy on:
- first blend epoch
- observed blend ρ
- off-class vote mass
- frac_blended
- allocation/merge rate
- memory occupancy
- accuracy / label purity
- whether real blend onset is no longer left-censored at epoch 0

## Scope
No `AgentAssociativeMemory` refactor yet. This issue is calibration only.

The agent-memory wrapper will inherit the same miscalibration if the write gate remains tied to non-transferable absolute ρ assumptions. Relative vigilance calibration is the prerequisite.

**Comment by samscarrow** [comment]
## Δ-relative vigilance: calibration result — does NOT fix epoch-0 left-censoring

**Headline (negative):** `RelativeVigilance` does **not** resolve the epoch-0 left-censoring of the blend onset on real embeddings. Margin and relative policies produce a **byte-identical** blend onset (ρ=0.233 @ epoch 0) on the real 20NG + all-MiniLM-L6-v2 manifold. The relative anchor is a partial fix, not the whole fix.

Ran `benchmarks/probe_contraction.py`, both modes × both policies, seed 0, 30 epochs:
```bash
# synthetic baseline
python benchmarks/probe_contraction.py --vigilance-policy {margin,relative} \
    --epochs 30 --classes 8 --dim 64 --contraction-end 0.97 --seed 0
# real transfer (sci.space, rec.sport.hockey, talk.politics.guns, comp.graphics; attractor sci.med)
python benchmarks/probe_contraction.py --real --vigilance-policy {margin,relative} \
    --epochs 30 --contraction-end 0.9 --seed 0 --margin-guard 0.05 --percentile 0.95
```

### Why the write gate can't fix this
The blend condition is a **retrieval-side** property, not a write-side one. Blend onset is defined by off-class softmax vote mass on the held-out probe — a function of `inference_temp` and the **resting geometry of the embedding manifold**, which already places off-class neighbors inside the retrieval neighborhood at contraction=0 (Δ≈0.42, off-class ρ≈0.23, so the gap is small *before any drift*). `RelativeVigilance` only changes which writes are admitted/merged into prototypes; it cannot retroactively separate retrieval neighborhoods that the encoder itself collapsed. That is exactly why both policies yield the identical onset — the held-out retrieval path is untouched by the write-gate rule.

| metric | synth-margin | synth-relative | real-margin | real-relative |
|---|---|---|---|---|
| Δ (within-class, e0) | 0.722 | 0.722 | 0.422 | 0.422 |
| first blend epoch | 21 | 20 | **0** | **0** |
| observed blend ρ | 0.703 | 0.678 | **0.233** | **0.233** |
| left-censored at e0 | no | no | **yes** | **yes** |
| max ρ observed | 0.780 | 0.779 | 0.453 | 0.443 |
| chimera (ρ ≥ v_ceiling=0.95) | unreachable | unreachable | unreachable | unreachable |

Secondary findings: the **Δ-anchored** blend formula `Δ − temp·ln((1−ε)/ε)` transfers (offsets within ±1 inference-temp logit step: synth +0.091/+0.066, real −0.079); the **absolute** ~0.79 onset does not (real gap −0.557); the absolute chimera criterion v_ceiling=0.95 is unreachable on the real manifold (max ρ≈0.45). Re-confirms #71/#72.

### Where relative vigilance still helps (write path)
It is not useless — it improves the **write path** by tightening merge behavior and prototype hygiene. On the real manifold it adapts v_eff down to the compressed geometry (0.61 vs margin's 0.79), which reduces unsafe merges:

| | real-margin | real-relative |
|---|---|---|
| final occupancy (slots) | 377 | **231** (−39%) |
| final accuracy (post-drift) | 0.375 | **0.414** (+0.039) |
| mean v_eff (e0 → e29) | 0.85 → 0.79 | 0.85 → **0.61** |

(On synthetic, margin saturates the full 4096-slot capacity; relative holds at 3237.) So relative vigilance is worth keeping as the write-gate policy — leaner memory, better degradation under drift — but it does not address the retrieval-side failure mode this issue's evaluation foregrounded.

### Follow-ups
1. **Instrumentation PR** — fold `occupied`, `merge_rate`, and held-out `acc` columns into `probe_contraction.py` so the write-side win is reproducible from `main` (the occupancy/accuracy rows above came from a supplementary script reusing the harness's batch sources, not the committed driver).
2. **Separate issue: retrieval-side de-censoring** — confidence-weighted retrieval / adaptive retrieval floor (live-Δ `RetrievalFloorPolicy`, #74) and a manifold-relative replacement for the unreachable absolute chimera criterion. This is the part that can actually move the epoch-0 onset; the write gate cannot.

<sub>seed 0, CPU, torch 2.8.0 / sentence-transformers 5.1.0.</sub>

**Comment by samscarrow** [comment]
Closing — the issue's two deliverables both shipped:

- **RelativeVigilance implementation + evaluation** — landed in the four-run calibration ([result](https://github.com/git-scarrow/fast-associative-memory/issues/73#issuecomment-4559126243)). Improves the write path (−39% occupancy, +0.04 accuracy on real 20NG) but does not move retrieval-side blend onset off epoch 0.
- **Instrumentation PR** — merged in #81, with the merge/alloc counts later corrected (PR review) to derive from `learn_local`'s literal hits/misses + actual allocation rather than the occupancy-delta proxy that mis-attributed eviction reuse as merging.

The retrieval-side de-censoring follow-up was correctly spun out to #80, and #80's central question is now answered (live-Δ floor moves real-20NG blend onset off epoch 0). Nothing left in this issue's scope.
